com.microsoft.java.lsif.core/src/com/microsoft/java/lsif/core/internal/visitors/LsifVisitor.java

```diff
@@ -106,12 +193,144 @@ private void resolve(int startPosition, int length, boolean needResolveImpl) {
 		}
 	}
 
-	private String createSymbolKey(Location definitionLocation) {
-		String rawKeyString = definitionLocation.toString();
-		return DigestUtils.md5Hex(rawKeyString);
+	private void resolveMoniker(IJavaElement element, MonikerKind monikerKind, SymbolData symbolData, Location definitionLocation, Range sourceRange) throws JavaModelException {
+		LsifService lsif = this.getLsif();
+		// Import Monikers
+		IJavaProject javaProject = element.getJavaProject();
+		IClassFile cf = (IClassFile) element.getAncestor(IJavaElement.CLASS_FILE);
+		PackageManager manager = null;
+		String packageName = "";
+		String version = "";
+		String type = "";
+		String url = "";
+		if (cf != null) {
+			IPath path = cf.getPath();
+			IPackageFragmentRoot root = javaProject.findPackageFragmentRoot(path);
+			IClasspathEntry container = root.getRawClasspathEntry();
+			IPath containerPath = container.getPath();
+			String pathName = containerPath.toString();
+			if (pathName.startsWith(JavaRuntime.JRE_CONTAINER)) {
+				// JDK Library
+				manager = PackageManager.JDK;
+				Manifest manifest = new Manifest();
+				if (root instanceof JarPackageFragmentRoot) {
+					manifest = ((JarPackageFragmentRoot) root).getManifest();
+					if (manifest != null) {
+						Attributes attributes = manifest.getMainAttributes();
+						version = attributes.getValue("Implementation-Version");
+					}
+				}
+				PackageFragmentRoot packageFragmentRoot = (PackageFragmentRoot) cf
+						.getAncestor(IJavaElement.PACKAGE_FRAGMENT_ROOT);
+				if (packageFragmentRoot != null) {
+					IModuleDescription moduleDescription = packageFragmentRoot.getAutomaticModuleDescription();
+					packageName = moduleDescription.getElementName();
+				}
+			} else {
+				manager = PackageManager.MAVEN;
+				MavenProject mavenProject = Repository.getInstance().enlistMavenProject(lsif, path);
+				if (mavenProject != null) {
+					Model model = mavenProject.getModel();
+					String groupId = model.getGroupId();
+					String artifactId = model.getArtifactId();
+					if (StringUtils.isNotEmpty(groupId) && StringUtils.isNotEmpty(artifactId)) {
+						packageName = groupId + "/" + artifactId;
+					}
+					version = model.getVersion();
+					Scm scm = model.getScm();
+					if (scm != null) {
+						url = scm.getUrl();
+						type = ScmUrlUtils.getProvider(scm.getConnection());
+					}
+				}
+			}
+		}
+
+		// Export Monikers
+		if (monikerKind == MonikerKind.EXPORT && this.isPublish) {
+			manager = PackageManager.MAVEN;
+		}
+
+		String identifier = "";
+		try {
```

Comment:
If we want to leave the identifier empty when exception happens, please move the error handling into the `getMonikerIdentifier()` and return "" when exception happens